kernel/src/engine/ensure_data_types.rs

```diff
@@ -0,0 +1,486 @@
+//! Helpers to ensure that kernel data types match arrow data types
+
+use std::collections::{HashMap, HashSet};
+
+use arrow_schema::{DataType as ArrowDataType, Field as ArrowField};
+use itertools::Itertools;
+
+use crate::{
+    engine::arrow_utils::make_arrow_error,
+    schema::{DataType, MetadataValue, StructField},
+    utils::require,
+    DeltaResult, Error,
+};
+
+/// Ensure a kernel data type matches an arrow data type. This only ensures that the actual "type"
+/// is the same, but does so recursively into structs, and ensures lists and maps have the correct
+/// associated types as well.
+///
+/// If `check_nullability_and_metadata` is true, this will also return an error if it finds a struct
+/// field that differs in nullability or metadata between the kernel and arrow schema. If it is
+/// false, no checks on nullability or metadata are performed.
+///
+/// This returns an `Ok(DataTypeCompat)` if the types are compatible, and
+/// will indicate what kind of compatibility they have, or an error if the types do not match. If
+/// there is a `struct` type included, we only ensure that the named fields that the kernel is
+/// asking for exist, and that for those fields the types match. Un-selected fields are ignored.
+pub(crate) fn ensure_data_types(
+    kernel_type: &DataType,
+    arrow_type: &ArrowDataType,
+    check_nullability_and_metadata: bool,
+) -> DeltaResult<DataTypeCompat> {
+    let check = EnsureDataTypes { check_nullability_and_metadata };
+    check.ensure_data_types(kernel_type, arrow_type)
+}
+
+struct EnsureDataTypes {
+    check_nullability_and_metadata: bool,
+}
+
+/// Capture the compatibility between two data-types, as passed to [`ensure_data_types`]
+pub(crate) enum DataTypeCompat {
+    /// The two types are the same
+    Identical,
+    /// What is read from parquet needs to be cast to the associated type
+    NeedsCast(ArrowDataType),
+    /// Types are compatible, but are nested types. This is used when comparing types where casting
+    /// is not desired (i.e. in the expression evaluator)
+    Nested,
+}
+
+impl EnsureDataTypes {
+    // Perform the check. See documentation for `ensure_data_types` entry point method above
+    fn ensure_data_types(
+        &self,
+        kernel_type: &DataType,
+        arrow_type: &ArrowDataType,
+    ) -> DeltaResult<DataTypeCompat> {
+        match (kernel_type, arrow_type) {
+            (DataType::Primitive(_), _) if arrow_type.is_primitive() => {
+                check_cast_compat(kernel_type.try_into()?, arrow_type)
+            }
+            (&DataType::BOOLEAN, ArrowDataType::Boolean)
+                | (&DataType::STRING, ArrowDataType::Utf8)
+                | (&DataType::BINARY, ArrowDataType::Binary) => {
+                    // strings, bools, and binary  aren't primitive in arrow
```

Comment:
nit: Maybe easier to understand if the comment comes first?
```suggestion
            // strings, bools, and binary  aren't primitive in arrow
            (&DataType::BOOLEAN, ArrowDataType::Boolean)
                | (&DataType::STRING, ArrowDataType::Utf8)
                | (&DataType::BINARY, ArrowDataType::Binary) => {
```